File: LeetCodeTasks/61_rotate-list/solutions/python3/solution.py
from typing import Optional

class Solution:

    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if not head or not head.next or k == 0:

            return head

        

        # First determine the length of the linked list

        length = 1

        current = head

        while current.next:

            current = current.next

            length += 1

        

        # Make the list circular

        current.next = head

        

        # Calculate the effective rotations needed

        k = k % length

        steps_to_new_head = length - k

        

        # Find the new head

        new_tail = head

        for _ in range(steps_to_new_head - 1):

            new_tail = new_tail.next

        new_head = new_tail.next

        

        # Break the circle

        new_tail.next = None

        

        return new_head

File: LeetCodeTasks/61_rotate-list/solutions/c/solution.c
#include <stddef.h>


struct ListNode* rotateRight(struct ListNode* head, int k) {

    if (head == NULL || head->next == NULL || k == 0) return head;



    // Compute the length of the list

    struct ListNode *current = head;

    int length = 1;

    while (current->next != NULL) {

        current = current->next;

        length++;

    }



    // Make the list circular

    current->next = head;



    // Calculate effective rotations needed

    k = k % length;

    int stepsToNewHead = length - k;

    

    // Find the new end of the rotated list

    struct ListNode *newTail = head;

    for (int i = 0; i < stepsToNewHead - 1; i++) {

        newTail = newTail->next;

    }



    // Set the new head

    struct ListNode *newHead = newTail->next;



    // Break the loop

    newTail->next = NULL;



    return newHead;

}

